### ISSUE TITLE
kv: limit size of range split keys

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#82543]
We've seen in support issues that range descriptors with large start and end keys can cause serious issues. RangeDescriptors are stored in memory, in the meta range keys, in range local keys, in the txn record performing range operations, and in the rangelog. As a result, large range descriptors can cause OOMs, large Raft writes, and other forms of issues.

We should place a reasonable limit on the size of range split keys, probably in `adminSplitWithDescriptor`. A limit like 4MB would prevent large data keys from leaking into metadata.

If we make this change, we'll need to carefully consider the effect that blocking splits can have on range size backpressure. If a range only has large keys and it reaches the backpressure size threshold, what should happen?

**Comment by petermattis** [comment]
Does the idea sketched out in https://github.com/cockroachdb/pebble/issues/1741#issuecomment-1148998051 apply to range split keys as well? Do we need range split keys to actually be real keys, or can they be synthesized as separators between real keys.

**Comment by nvanbenschoten** [comment]
I was thinking something similar. A range's start key does not need to be a real user key, so we could truncate the keys in a range to some maximum size when computing a range split key. That would work in many cases, though it still does run into issues when many keys share the same pre-truncated prefix. At the extreme where all keys in a range share the same prefix (up to the truncation size), there is no valid key to split at while adhering to a key size limit.

**Comment by petermattis** [comment]
> At the extreme where all keys in a range share the same prefix (up to the truncation size), there is no valid key to split at while adhering to a key size limit.

True, though I would expect this to be very rare. Certainly, this scenario should be protected against via other mechanisms (e.g. a max size on range split keys).
